src/repr/type_random.ml

```diff
@@ -0,0 +1,138 @@
+open Type_core
+open Staging
+module R = Random.State
+
+type 'a random = (R.t -> 'a) staged
+
+let ( let+ ) x f =
+  let x = unstage x in
+  stage (f x)
+
+let ( and+ ) a b = stage (unstage a, unstage b)
+let return x = stage (fun _ -> x)
+
+(* Sample lengths according to a geometric distribution by inverse
+   transform sampling an exponential one (characterised by [mean_length = λ⁻¹])
+   and rounding to integer values. *)
+let pick_len : mean:int -> len -> R.t -> int =
+  let bound len x =
+    match len with
+    | `Int | `Int64 -> x
+    | `Int8 -> min x ((1 lsl 8) - 1)
+    | `Int16 -> min x ((1 lsl 16) - 1)
+    | `Int32 -> min x ((0x7fff lsl 16) lor 0xffff)
+  in
+  fun ~mean l s ->
+    match l with
+    | (`Int | `Int8 | `Int16 | `Int32 | `Int64) as l ->
+        bound l (Float.to_int (-.Float.log (R.float s 1.) *. Float.of_int mean))
+    | `Fixed i -> i
+
+let indexable :
+    type a b.
+    mean_len:int -> len -> (int -> (int -> a) -> b) -> a random -> b random =
+ fun ~mean_len len init elt ->
+  let+ elt = elt in
+  fun s -> init (pick_len ~mean:mean_len len s) (fun _ -> elt s)
+
+module Record_deriver = Fields_folder (struct
+  type ('a, 'b) t = R.t -> 'b -> 'a
+end)
+
+let int32 =
+  let open Int32 in
+  let bits s = of_int (R.bits s) in
+  fun s -> logxor (bits s) (shift_left (bits s) 30)
+
+let int64 =
+  let open Int64 in
+  let bits s = of_int (R.bits s) in
+  fun s ->
+    logxor (bits s) (logxor (shift_left (bits s) 30) (shift_left (bits s) 60))
+
+let int =
+  match Sys.word_size with
+  | 64 -> fun s -> Int64.to_int (int64 s)
+  | 32 -> fun s -> Int32.to_int (int32 s)
+  | _ -> assert false
+
+let float s =
+  R.float s (if R.bool s then Float.max_float else -.Float.max_float)
+
+let rec t : type a. a t -> a random = function
+  | Map x -> map x
+  | Prim x -> prim x
+  | Tuple x -> tuple x
+  | List { len; v } -> indexable ~mean_len:4 len List.init (t v)
+  | Array { len; v } -> indexable ~mean_len:4 len Array.init (t v)
+  | Option x -> option x
+  | Record x -> record x
+  | Variant x -> variant x
+  | Boxed x -> t x
+  | Self x -> stage (fun s -> (* improperly staged *) unstage (t x.self_fix) s)
+  | Custom _ -> failwith "Cannot generate random instance of Custom type"
+  | Var v -> raise (Unbound_type_variable v)
+
+and char : char random = stage (fun s -> Char.unsafe_chr (R.int s 256))
+
+and prim : type a. a prim -> a random = function
+  | Unit -> return ()
+  | Bool -> stage R.bool
+  | Char -> char
+  | Int -> stage int
+  | Int32 -> stage int32
+  | Int64 -> stage int64
+  | Float -> stage float
+  | String len -> indexable ~mean_len:8 len String.init char
+  | Bytes len -> indexable ~mean_len:8 len Bytes.init char
```

Comment:
Shouldn't `~mean_len` be an optional parameter of this function and the function `t` above? (even though I remember you don't like optional parameters)

Reply:
I'm not sure there's a sensible default for `mean_len` here: too low and we get lots of collisions in samples, too high and we risk generating huge samples (esp. for nested container types). (And yes, I'm definitely not a fan of optional parameters for internal functions -- I've written a lot of bugs that way.)

Reply:
The thing is you provided a default value by writing `~mean_len:8` and I can't find a way to manage this value from a user perspective, no?

Reply:
Ah, I see. Yes, it's not possible to change this behaviour at the moment. (Likewise for the choice of 4 for lists and arrays.)

I think it makes sense to allow these four hyperparams to be modified via typerep metadata, as we do for the other generics, but currently this is only possible by extending `Custom`. I'd like to change `Custom` to make this metadata extensible, since it's getting quite unweildy. Adding parameters to the generic operation itself is more brittle since those settings affect all lists / strings etc. inside the typerep (and in principle a consumer doesn't know where all of those are), which is probably why we tend towards the former solution.

Reply:
Ok, that's fine for me! Thanks :-)